pkg/pgmodel/sql_ingest.go

```diff
@@ -721,42 +764,104 @@ func (h *insertHandler) setSeriesIds(sampleInfos []samplesInfo) (string, error)
 
 	br, err := h.conn.SendBatch(context.Background(), batch)
 	if err != nil {
-		return "", err
+		return "", epoch, err
 	}
 	defer br.Close()
 
 	if numSQLFunctionCalls != len(batchSeries) {
-		return "", fmt.Errorf("unexpected difference in numQueries and batchSeries")
+		return "", epoch, fmt.Errorf("unexpected difference in numQueries and batchSeries")
 	}
 
 	var tableName string
 	for i := 0; i < numSQLFunctionCalls; i++ {
 		_, err = br.Exec()
 		if err != nil {
-			return "", err
+			return "", epoch, err
 		}
+
+		var newEpoch int64
 		row := br.QueryRow()
+		err = row.Scan(&newEpoch)
+		if err != nil {
+			return "", epoch, err
+		}
+		if newEpoch < epoch {
+			// it's always safe to stay on an old epoch, so we don't update
+			// the epoch for the cache until we're sure all the queries succeeded
+			epoch = newEpoch
+		}
 
 		var id SeriesID
+		row = br.QueryRow()
 		err = row.Scan(&tableName, &id)
 		if err != nil {
-			return "", err
+			return "", epoch, err
 		}
-		h.seriesCache[batchSeries[i][0].labels.String()] = id
+		seriesString := batchSeries[i][0].labels.String()
+		h.seriesCache[seriesString] = id
 		for _, lsi := range batchSeries[i] {
 			lsi.seriesID = id
 		}
+
 		_, err = br.Exec()
 		if err != nil {
-			return "", err
+			return "", epoch, err
 		}
 	}
 
-	return tableName, nil
+	return tableName, epoch, nil
+}
+
+func (h *insertHandler) refreshSeriesCache() {
+	newEpoch, err := h.getServerEpoch()
+	if err != nil {
+		// we don't have any great place to report this error, and if the
+		// connection recovers we can still make progress, so we'll just log it
+		// and continue execution
+		msg := fmt.Sprintf("error refreshing the series cache for %s", h.metricTableName)
+		log.Error("msg", msg, "err", err)
+		// Trash the cache just in case an epoch change occured, seems safer
+		h.seriesCache = map[string]SeriesID{}
+		return
+	}
+
+	if newEpoch != h.seriesCacheEpoch {
+		h.seriesCache = map[string]SeriesID{}
+		h.seriesCacheEpoch = newEpoch
+	}
+}
+
+func (h *insertHandler) getServerEpoch() (Epoch, error) {
+	// using a batch to get QueryRow
+	batch := h.conn.NewBatch()
```

Comment:
This does not need to be a batch.

Reply:
yeah, I'm just sing it to get `QueryRow` (halves the size of this function)